File: app/src/main/java/com/example/camera/activity/CameraActivity.java
package com.example.camera.activity;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.hardware.Camera;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.util.Size;
import android.view.KeyEvent;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.webkit.MimeTypeMap;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.example.camera.R;
import com.example.camera.base.BaseActivity;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.Executor;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.camera.core.CameraX;
import androidx.camera.core.FlashMode;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageAnalysisConfig;
import androidx.camera.core.ImageCapture;
import androidx.camera.core.ImageCaptureConfig;
import androidx.camera.core.ImageProxy;
import androidx.camera.core.Preview;
import androidx.camera.core.PreviewConfig;
import androidx.lifecycle.LifecycleOwner;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * @author libo
 * @Demo class CameraActivity
 * @Description TODO
 * @date 2019-11-14 15:33
 */
public class CameraActivity extends BaseActivity {

	private static final String TAG = "CameraActivity";

	@BindView(R.id.tv_camera)
	TextureView tvCamera;
	@BindView(R.id.rl_content)
	RelativeLayout rlContent;
	@BindView(R.id.iv_light)
	ImageView ivLight;
	@BindView(R.id.iv_change)
	ImageView ivChange;
	@BindView(R.id.iv_camera)
	ImageView ivCamera;
	@BindView(R.id.iv_gallery)
	ImageView ivGallery;

	private Preview preview;
	private ViewGroup parent;

	private Matrix matrix = new Matrix();
	private int centerX;
	private int centerY;
	private int rotationDegrees;
	private ImageCapture imageCapture;

	private File file;
	private File fileDir;
	private CameraX.LensFacing lensFacing;

	private boolean isOnLight = false;

	@Override
	protected void onCreateView() {
		getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
		setContentView(R.layout.activity_camera);

		file = new File(Environment.getExternalStorageDirectory() + "/相机拍照/" + System.currentTimeMillis() + ".jpg");
		fileDir = new File(Environment.getExternalStorageDirectory() + "/相机拍照/");
		if (!fileDir.exists()) {
			fileDir.mkdirs();
			if (!file.exists()) {
				try {
					file.createNewFile();
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
		}
	}

	@Override
	protected void initView(@Nullable Bundle savedInstanceState) {
		PreviewConfig previewConfig = new PreviewConfig.Builder().setTargetResolution(new Size(1080, 2160)).setTargetRotation(getWindowManager().getDefaultDisplay().getRotation()).build();
		preview = new Preview(previewConfig);
		preview.setOnPreviewOutputUpdateListener(output -> {
			parent = (ViewGroup) tvCamera.getParent();
			parent.removeView(tvCamera);
			parent.addView(tvCamera, 0);
			tvCamera.setSurfaceTexture(output.getSurfaceTexture());
			updateTransform();
		});
		ImageCaptureConfig config = new ImageCaptureConfig.Builder().setCaptureMode(ImageCapture.CaptureMode.MIN_LATENCY).setTargetResolution(new Size(1280, 720)).build();
		imageCapture = new ImageCapture(config);
		CameraX.bindToLifecycle(this, preview, imageCapture);
	}

	private void updateTransform() {
		centerX = tvCamera.getWidth() / 2;
		centerY = tvCamera.getHeight() / 2;
		switch (getWindowManager().getDefaultDisplay().getRotation()) {
			case Surface.ROTATION_0:
				rotationDegrees = 0;
				break;
			case Surface.ROTATION_90:
				rotationDegrees = 90;
				break;
			case Surface.ROTATION_180:
				rotationDegrees = 180;
				break;
			case Surface.ROTATION_270:
				rotationDegrees = 270;
				break;
		}
		matrix.postRotate(-rotationDegrees, centerX, centerY);
		tvCamera.setTransform(matrix);
	}

	@OnClick({R.id.iv_light, R.id.iv_change, R.id.iv_camera, R.id.iv_gallery})
	public void onViewClicked(View view) {
		switch (view.getId()) {
			case R.id.iv_light:
				isOnLight = !isOnLight;
				ivLight.setImageResource(isOnLight ? R.drawable.icon_light_on : R.drawable.icon_light_off);
				// AUTO 自动闪光   OFF 不闪光   ON 一直闪烁
				imageCapture.setFlashMode(isOnLight ? FlashMode.AUTO : FlashMode.OFF);
				break;
			case R.id.iv_change:
				lensFacing = CameraX.LensFacing.FRONT == lensFacing ? CameraX.LensFacing.BACK : CameraX.LensFacing.FRONT;
				break;
			case R.id.iv_camera:
				startCamera();
				break;
			case R.id.iv_gallery:
				break;
		}
	}

	/**
	 * 拍照
	 */
	private void startCamera() {
		imageCapture.takePicture(file, command -> {

		}, new ImageCapture.OnImageSavedListener() {
			@Override
			public void onImageSaved(@NonNull File file) {
				uodateGallery(file);
				if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
					sendBroadcast(new Intent(Camera.ACTION_NEW_PICTURE, Uri.fromFile(file)));
				}
//				 mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(file.extension);
//				MediaScannerConnection.scanFile(context, file.absolutePath, mimeType, null);
			}

			@Override
			public void onError(@NonNull ImageCapture.ImageCaptureError imageCaptureError, @NonNull String message, @Nullable Throwable cause) {

			}
		});
	}

	private void uodateGallery(File file) {
		ivGallery.post(() -> {
			Glide.with(ivGallery).load(file).apply(RequestOptions.circleCropTransform()).into(ivGallery);
		});
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		CameraX.setErrorListener((error, message) -> {
			Log.e(TAG, message);
		}, null);
		CameraX.unbind(preview, imageCapture);
		preview.removePreviewOutputListener();
	}
}

File: KotlinCameraX/src/main/java/com/example/camerax/activity/CameraActivity.kt
package com.example.camerax.activity

import android.content.Intent
import android.hardware.Camera
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.util.*
import android.view.ViewGroup
import androidx.camera.core.*
import com.bumptech.glide.Glide
import com.example.camerax.R
import com.example.camerax.base.BaseActivity
import com.example.camerax.tools.LuminosityAnalyzer
import kotlinx.android.synthetic.main.activity_camera.*
import java.io.File
import java.util.concurrent.Executor

/**
 * @Demo class CameraActivity
 * @Description TODO
 * @author libo
 * @date 2019-11-19 11:31
 */
class CameraActivity : BaseActivity() {

    private lateinit var imageCapture: ImageCapture
    private lateinit var file: File
    private var lensFacing = CameraX.LensFacing.BACK
    private var flashMode = FlashMode.AUTO

    override fun onCreateView() {
        setContentView(R.layout.activity_camera)
    }

    override fun initView(savedInstanceState: Bundle?) {
        ttVCamera.post { initPreview() }
    }

    fun initPreview() {
        CameraX.unbindAll()
        val metrics = DisplayMetrics().also { ttVCamera.display.getRealMetrics(it) }
        val screenSize = Size(metrics.widthPixels, metrics.heightPixels)
        //预览
        val previewConfig = PreviewConfig.Builder()
                .setLensFacing(lensFacing)
                .setTargetResolution(screenSize)
                .build()

        val preview = Preview(previewConfig)

        preview.setOnPreviewOutputUpdateListener {
            val parent = ttVCamera.parent as ViewGroup
            parent.removeView(ttVCamera)
            parent.addView(ttVCamera, 0)
            ttVCamera.surfaceTexture = it.surfaceTexture
        }

        //图像分析
        val analyzerConfig = ImageAnalysisConfig.Builder().apply {
            setLensFacing(lensFacing)
            setImageReaderMode(ImageAnalysis.ImageReaderMode.ACQUIRE_LATEST_IMAGE)
        }.build()

        //拍照
        val captureConfig = ImageCaptureConfig.Builder()
                .setLensFacing(lensFacing)
                .setCaptureMode(ImageCapture.CaptureMode.MIN_LATENCY)
                .setTargetRotation(ttVCamera.display.rotation)
                .build()

        imageCapture = ImageCapture(captureConfig)
        imageCapture.flashMode = flashMode

        val analyzerUseCase = ImageAnalysis(analyzerConfig).apply {
            setAnalyzer(Executor {
                Log.i("CameraActivity", "aaaaaaaaaaaaa  图片美化")
            }, LuminosityAnalyzer())
        }
        //绑定生命周期
        CameraX.bindToLifecycle(this, preview, analyzerUseCase, imageCapture)
        setListener()
    }

    private fun setListener() {
        ivCamera.setOnClickListener {
            file = File(Environment.getExternalStorageDirectory().toString() + "/相机拍照/" + System.currentTimeMillis() + ".jpg")
            imageCapture.takePicture(file, {
                Log.i("CameraActivity", "aaaaaaaaaaaaa  图片压缩")
            }, object : ImageCapture.OnImageSavedListener {
                override fun onError(imageCaptureError: ImageCapture.ImageCaptureError, message: String, exc: Throwable?) {

                }

                override fun onImageSaved(file: File) {
                    Glide.with(context).load(file).into(ivGallery)
                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
                        sendBroadcast(Intent(Camera.ACTION_NEW_PICTURE, Uri.fromFile(file)))
                    }
                }
            })
        }

        ivLight.setOnClickListener {
            // AUTO 自动闪光   OFF 不闪光   ON 一直闪烁
            flashMode = if (flashMode == FlashMode.OFF) FlashMode.AUTO else FlashMode.OFF
            ivLight.setImageResource(if (flashMode == FlashMode.AUTO) R.drawable.icon_light_on else R.drawable.icon_light_off)
            initPreview()
        }

        ivChange.setOnClickListener {
            lensFacing = if(CameraX.LensFacing.FRONT == lensFacing) CameraX.LensFacing.BACK else CameraX.LensFacing.FRONT
            initPreview()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        CameraX.unbindAll()
    }
}